plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/inline/InlineChatController.kt

```diff
@@ -280,112 +282,213 @@ class InlineChatController(
-                    addedCharsCount += row.newLine?.length ?: 0
+                    numSuggestionAddLines += 1
+                    numSuggestionAddChars += row.newLine.length
                 }
             }
         }
-        metrics?.numSuggestionAddChars = addedCharsCount
-        metrics?.numSuggestionAddLines = addedLinesCount
-        metrics?.numSuggestionDelChars = deletedCharsCount
-        metrics?.numSuggestionDelLines = deletedLinesCount
-        return isAllEqual
+        metrics?.numSuggestionAddChars = numSuggestionAddChars
+        metrics?.numSuggestionAddLines = numSuggestionAddLines
+        metrics?.numSuggestionDelChars = numSuggestionDelChars
+        metrics?.numSuggestionDelLines = numSuggestionDelLines
+        if (isAllEqual) {
+            canPopupAbort.set(true)
+            throw Exception("No suggestions from Q; please try a different instruction.")
+        }
     }
 
     private fun processChatDiff(selectedCode: String, event: ChatMessage, editor: Editor, selectionRange: RangeMarker) {
+        if (isAbandoned.get()) return
         if (event.message?.isNotEmpty() == true) {
+            logger.info { "inline chat recommendation: \n ${event.message}" }
             runBlocking {
-                while (partialUndoActions.isNotEmpty()) {
-                    val action = partialUndoActions.pop()
-                    runBlocking { action.invoke() }
-                }
-                partialAcceptActions.clear()
-
                 val recommendation = unescape(event.message)
-                if (selectedCode == recommendation) {
-                    throw Exception("No suggestions from Q; please try a different instruction.")
-                }
                 val selection = selectedCode.split("\n")
                 val recommendationList = recommendation.split("\n")
                 val diff = compareDiffs(selection, recommendationList)
-
+                val startLine = getLineNumber(editor.document, selectionRange.startOffset)
+                val endLine = getLineNumber(editor.document, selectionRange.endOffset)
+                val patchString = constructPatch(diff)
                 if (currentPopup?.isVisible != true) {
                     logger.debug { "inline chat popup cancelled before diff is shown" }
                     isInProgress.set(false)
+                    isAbandoned.set(true)
                     recordInlineChatTelemetry(InlineChatUserDecision.DISMISS)
                     return@runBlocking
                 }
-
-                val isAllEqual = processDiffRows(diff, selectionRange, editor)
-                if (isAllEqual) {
-                    throw Exception("No suggestions from Q; please try a different instruction.")
+                launch(EDT) {
+                    removeSelection(editor)
+                    applyChunk(patchString, editor, startLine, endLine)
+                    processHighlights(diff, startLine, editor)
+                }.join()
```

Comment:
same comment as above

Reply:
unclear what the delta is between this and the above method

Reply:
there are common part can be extracted out between the two, I will take it as a followup item